### ISSUE TITLE
When an update action incurs validation errors, Devise always redirects to default #edit_user_registration_path

### ISSUE THREAD

**Issue by dannguyen** [plataformatec/devise#2788]
For Rails **3.2.15** and Devise **3.2.2**

I've broken up my `user` editing view into two different views: user information (email, name, password) and billing information, with the former using the standard "devise/registrations/edit" template and route, and the latter using a route and action that I've named `/edit_billing`

In this `edit_billing` action, I require the `current_password` to be  re-entered. The action that `edit_billing` submits is to a custom `update` that I've named `update_billing` (via `:puts`). It acts similar to the standard `update` in DeviseController::Registrations: the user is sent to their `users/show` page, or to any page I'd like, via the standard `respond_with` call. Here's my code:

```
def update_billing
    user = current_user
    stripe_token = params[:user][:stripe_token]
    current_password = params[:user][:current_password]    

    if update_resource(user, account_update_params)
      # Call method to update billing info stored on Stripe's server
      user.update_billing_from_stripe_with_token(stripe_token)
      sign_in resource_name, user, :bypass => true
      respond_with user, location: '/profile'
    else
      clean_up_passwords user
      respond_with user, :location => "/edit_billing" 
    end    
end
```

But when `current_password` is not valid -- or any other validation method -- fails, the `user` resource incurs errors as expected. **However**, `respond_with` does not obey the specified `location` parameter, and will always redirect the user to the `devise/registrations/edit` page, where it will show the correct Devise errors, but of course, I prefer the user to see those errors on the `edit_billing` route.

Part of this seems to be surprising behavior in the Rails `respond_with` call, at least in 3.2.15. This blog post sums it up well:

http://bendyworks.com/geekville/tutorials/2012/6/respond-with-an-explanation

> This is why redefining after_sign_up_path_for in your controller allows you to change the post-signup location. Cool, right? Only … this wasn’t working. To understand why it wasn’t working, we needed to dig into how respond_with works.
> 
> As always, Ryan Bates explains it best. Paraphrasing the relevant information: the most important thing to know is that respond_with is specific to an HTTP verb. If it’s a GET, it first looks for a view for that particular action and MIME type. If it can’t find that, it essentially calls #to_<format> on the object you give it. However, if it’s a POST, it will validate the object (and send it back to the appropriate form, if necessary), and then basically call redirect_to @object.
> 
> However, that’s not quite the whole story: respond_with ALWAYS looks for a corresponding template for an action, regardless of verb and MIME type, and regardless of whether a :location is passed to it or not (this behavior isn’t really documented anywhere, at least that we could find).

So I'm not sure if the behavior I'm experiencing is a result of this Rails behavior or if there is a Devise callback that I'm unaware of that is impacting the intended direct. 

To sum up, when the user object has `errors.present?`, the call to:

```
respond_with user, :location => "/edit_billing" 
```

Does _not_ go to the intended location, but to the default `devise/registrations/edit` view.

And when I call this:

```
respond_with user, :action => "edit_billing" 
```

The proper template **is** shown. However, the user object no longer available and so no errors are shown (and the form input is lost).

However, without errors, redirection/respond_with works as expected
